File: apps/mobile-legacy-rn/src/__tests__/simulation.test.ts
import { SimulationEngine } from '@/services/simulation/SimulationEngine';
import { demoSimulationPath, demoStops } from '@/mocks/demoData';

jest.useFakeTimers();

describe('SimulationEngine', () => {
  it('başlangıçta ilk konumu yayınlar', () => {
    const engine = new SimulationEngine({ path: demoSimulationPath, stops: demoStops, intervalMs: 1000 });
    const ticks: number[] = [];
    engine.subscribe((t) => ticks.push(t.location.latitude));
    expect(ticks).toHaveLength(1);
  });

  it('adım ilerledikçe araç konumu değişir', () => {
    const engine = new SimulationEngine({ path: demoSimulationPath, stops: demoStops, intervalMs: 1000 });
    let last = { latitude: 0, longitude: 0 };
    engine.subscribe((t) => {
      last = t.location;
    });
    const first = { ...last };
    engine.start();
    jest.advanceTimersByTime(3000);
    engine.stop();
    expect(last).not.toEqual(first);
  });

  it('yol sonunda finished=true olur', () => {
    const shortPath = demoSimulationPath.slice(0, 3);
    const engine = new SimulationEngine({ path: shortPath, stops: demoStops, intervalMs: 500 });
    let finished = false;
    engine.subscribe((t) => {
      finished = t.finished;
    });
    engine.start();
    jest.advanceTimersByTime(5000);
    expect(finished).toBe(true);
  });
});

File: apps/mobile/test/simulation_test.dart
import 'package:fake_async/fake_async.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:servis_takip/data/mock/demo_data.dart';
import 'package:servis_takip/data/simulation/simulation_engine.dart';

void main() {
  test('başlangıçta ilk konumu yayınlar', () {
    final engine = SimulationEngine(path: demoSimulationPath, stops: demoStops);
    final tick = engine.current;
    expect(tick.location, isNotNull);
    engine.dispose();
  });

  test('adım ilerledikçe araç konumu değişir', () {
    fakeAsync((async) {
      final engine = SimulationEngine(
        path: demoSimulationPath,
        stops: demoStops,
        interval: const Duration(seconds: 1),
      );
      final first = engine.current.location;
      engine.start();
      async.elapse(const Duration(seconds: 3));
      final later = engine.current.location;
      expect(later.latitude == first.latitude && later.longitude == first.longitude, isFalse);
      engine.dispose();
    });
  });

  test('yol sonunda finished=true olur', () {
    fakeAsync((async) {
      final shortPath = demoSimulationPath.sublist(0, 3);
      final engine = SimulationEngine(
        path: shortPath,
        stops: demoStops,
        interval: const Duration(milliseconds: 500),
      );
      var finished = false;
      engine.stream.listen((tick) => finished = tick.finished);
      engine.start();
      async.elapse(const Duration(seconds: 5));
      expect(finished, isTrue);
      engine.dispose();
    });
  });
}
